Understood, it was decided to deal with that later if it had to be dealt with at all anyway.

I have fixed my own tests, but the new implementation introduces a bunch of regressions that I'm digging into.

```
======================================================================
FAIL: test_escape_backslashes (test.test_email.test_email.TestMiscellaneous)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3080, in test_escape_backslashes
    self.assertEqual(utils.parseaddr(utils.formataddr((a, b))), (a, b))
AssertionError: Tuples differ: ('Arthur \\Backslash\\ Foobar', 'person@') != ('Arthur \\Backslash\\ Foobar', 'person@dom.ain')
First differing element 1:
'person@'
'person@dom.ain'
- ('Arthur \\Backslash\\ Foobar', 'person@')
+ ('Arthur \\Backslash\\ Foobar', 'person@dom.ain')
?                                         +++++++
======================================================================
FAIL: test_escape_dump (test.test_email.test_email.TestMiscellaneous)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3067, in test_escape_dump
    self.assertEqual(
AssertionError: Tuples differ: ('A (Very) Silly Person', 'person@') != ('A (Very) Silly Person', 'person@dom.ain')
First differing element 1:
'person@'
'person@dom.ain'
- ('A (Very) Silly Person', 'person@')
+ ('A (Very) Silly Person', 'person@dom.ain')
?                                   +++++++
======================================================================
FAIL: test_formataddr_does_not_quote_parens_in_quoted_string (test.test_email.test_email.TestMiscellaneous)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3172, in test_formataddr_does_not_quote_parens_in_quoted_string
    self.assertEqual(utils.parseaddr(addrstr), addr)
AssertionError: Tuples differ: ("'foo@example.com' (foo@example.com)", 'foo@') != ("'foo@example.com' (foo@example.com)", 'foo@example.com')
First differing element 1:
'foo@'
'foo@example.com'
- ("'foo@example.com' (foo@example.com)", 'foo@')
+ ("'foo@example.com' (foo@example.com)", 'foo@example.com')
?                                              +++++++++++
======================================================================
FAIL: test_getaddresses (test.test_email.test_email.TestMiscellaneous)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3209, in test_getaddresses
    eq(utils.getaddresses(['aperson@dom.ain (Al Person)',
AssertionError: Lists differ: [('Al Person', 'aperson@'), ('Bud Person', 'bperson@')] != [('Al Person', 'aperson@dom.ain'), ('Bud Person', 'bperson@dom.ain')]
First differing element 0:
('Al Person', 'aperson@')
('Al Person', 'aperson@dom.ain')
- [('Al Person', 'aperson@'), ('Bud Person', 'bperson@')]
+ [('Al Person', 'aperson@dom.ain'), ('Bud Person', 'bperson@dom.ain')]
?                         +++++++                            +++++++
======================================================================
FAIL: test_getaddresses_embedded_comment (test.test_email.test_email.TestMiscellaneous)
Test proper handling of a nested comment
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3228, in test_getaddresses_embedded_comment
    eq(addrs[0][1], 'foo@bar.com')
AssertionError: 'foo@' != 'foo@bar.com'
- foo@
+ foo@bar.com
======================================================================
FAIL: test_getaddresses_nasty (test.test_email.test_email.TestMiscellaneous)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3220, in test_getaddresses_nasty
    eq(utils.getaddresses(
AssertionError: Lists differ: [('', ''), ('Jason R. Mastaler', 'jason@')] != [('', ''), ('Jason R. Mastaler', 'jason@dom.ain')]
First differing element 1:
('Jason R. Mastaler', 'jason@')
('Jason R. Mastaler', 'jason@dom.ain')
- [('', ''), ('Jason R. Mastaler', 'jason@')]
+ [('', ''), ('Jason R. Mastaler', 'jason@dom.ain')]
?                                         +++++++
======================================================================
FAIL: test_multiline_from_comment (test.test_email.test_email.TestMiscellaneous)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3180, in test_multiline_from_comment
    self.assertEqual(utils.parseaddr(x), ('Foo Bar', 'foo@example.com'))
AssertionError: Tuples differ: ('Foo Bar', 'foo@') != ('Foo Bar', 'foo@example.com')
First differing element 1:
'foo@'
'foo@example.com'
- ('Foo Bar', 'foo@')
+ ('Foo Bar', 'foo@example.com')
?                  +++++++++++
======================================================================
FAIL: test_name_with_dot (test.test_email.test_email.TestMiscellaneous)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3126, in test_name_with_dot
    self.assertEqual(utils.parseaddr(x), (a, b))
AssertionError: Tuples differ: ('John X. Doe', 'jxd@') != ('John X. Doe', 'jxd@example.com')
First differing element 1:
'jxd@'
'jxd@example.com'
- ('John X. Doe', 'jxd@')
+ ('John X. Doe', 'jxd@example.com')
?                      +++++++++++
======================================================================
FAIL: test_parseaddr_multiple_domains (test.test_email.test_email.TestMiscellaneous)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3045, in test_parseaddr_multiple_domains
    self.assertEqual(
AssertionError: Tuples differ: ('', '') != ('', 'a@')
First differing element 1:
''
'a@'
- ('', '')
+ ('', 'a@')
?       ++
======================================================================
FAIL: test_parseaddr_preserves_quoted_pairs_in_addresses (test.test_email.test_email.TestMiscellaneous)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3142, in test_parseaddr_preserves_quoted_pairs_in_addresses
    eq(utils.parseaddr('""example" example"@example.com'),
AssertionError: Tuples differ: ('', '""example" example"@') != ('', '""example" example"@example.com')
First differing element 1:
'""example" example"@'
'""example" example"@example.com'
- ('', '""example" example"@')
+ ('', '""example" example"@example.com')
?                           +++++++++++
======================================================================
FAIL: test_parseaddr_preserves_spaces_in_local_part (test.test_email.test_email.TestMiscellaneous)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/test/test_email/test_email.py", line 3156, in test_parseaddr_preserves_spaces_in_local_part
    self.assertEqual(('', "merwok wok@xample.com"),
AssertionError: Tuples differ: ('', 'merwok wok@xample.com') != ('', 'merwok wok@')
First differing element 1:
'merwok wok@xample.com'
'merwok wok@'
- ('', 'merwok wok@xample.com')
?                  ----------
+ ('', 'merwok wok@')
```

--- a/Lib/email/_header_value_parser.py
+++ b/Lib/email/_header_value_parser.py
@@ -1587,6 +1587,8 @@ def get_domain(value):
         token, value = get_dot_atom(value)
     except errors.HeaderParseError:
         token, value = get_atom(value)
+    if value and value[0] == '@':
+        raise errors.HeaderParseError('Invalid Domain')
     if leader is not None:
         token[:0] = [leader]
     domain.append(token)

--- a/Lib/email/_parseaddr.py
+++ b/Lib/email/_parseaddr.py
@@ -379,7 +379,12 @@ class AddrlistClass:
         aslist.append('@')
         self.pos += 1
         self.gotonext()
-        return EMPTYSTRING.join(aslist) + self.getdomain()
+        domain = self.getdomain()
+        if not domain:
+            # Invalid domain, return an empty address instead of returning a
+            # local part to denote failed parsing.
+            return EMPTYSTRING
+        return EMPTYSTRING.join(aslist) + domain
 
     def getdomain(self):
         """Get the complete domain name from an address."""
@@ -394,6 +399,10 @@ class AddrlistClass:
             elif self.field[self.pos] == '.':
                 self.pos += 1
                 sdlist.append('.')
+            elif self.field[self.pos] == '@':
+                # bpo-34155: Don't parse domains with two `@` like
+                # `a@malicious.org@important.com`.
+                return EMPTYSTRING
             elif self.field[self.pos] in self.atomends:
                 break
             else:

--- a/Lib/test/test_email/test__header_value_parser.py
+++ b/Lib/test/test_email/test__header_value_parser.py
@@ -1444,6 +1444,16 @@ class TestParser(TestParserMixin, TestEmailBase):
         self.assertEqual(addr_spec.domain, 'example.com')
         self.assertEqual(addr_spec.addr_spec, 'star.a.star@example.com')
 
+    def test_get_addr_spec_multiple_domains(self):
+        with self.assertRaises(errors.HeaderParseError):
+            parser.get_addr_spec('star@a.star@example.com')
+
+        with self.assertRaises(errors.HeaderParseError):
+            parser.get_addr_spec('star@a@example.com')
+
+        with self.assertRaises(errors.HeaderParseError):
+            parser.get_addr_spec('star@172.17.0.1@example.com')
+
     # get_obs_route
 
     def test_get_obs_route_simple(self):

--- a/Lib/test/test_email/test_email.py
+++ b/Lib/test/test_email/test_email.py
@@ -3041,6 +3041,20 @@ class TestMiscellaneous(TestEmailBase):
         self.assertEqual(utils.parseaddr('<>'), ('', ''))
         self.assertEqual(utils.formataddr(utils.parseaddr('<>')), '')
 
+    def test_parseaddr_multiple_domains(self):
+        self.assertEqual(
+            utils.parseaddr('a@b@c'),
+            ('', '')
+        )
+        self.assertEqual(
+            utils.parseaddr('a@b.c@c'),
+            ('', '')
+        )
+        self.assertEqual(
+            utils.parseaddr('a@172.17.0.1@c'),
+            ('', '')
+        )
+
     def test_noquote_dump(self):
         self.assertEqual(
             utils.formataddr(('A Silly Person', 'person@dom.ain')),

--- a/Misc/NEWS.d/next/Security/2019-05-04-13-33-37.bpo-34155.MJll68.rst
+++ b/Misc/NEWS.d/next/Security/2019-05-04-13-33-37.bpo-34155.MJll68.rst
@@ -1,0 +1,1 @@
+Fix parsing of invalid email addresses with more than one ``@`` (e.g. a@b@c.com.) to not return the part before 2nd ``@`` as valid email address. Patch by maxking & jpic.

